Buffered deserialisation

JJ11teen opened the pull request:
Hello!

This is a draft PR to add support for deserialising yaml into different types, depending on values found during deserialisation. Specifically, it addresses scenarios mentioned in: #36 #343, #605, #731.

The foundational code included in this PR was written by @atruskie in a [gist available here]( https://gist.github.com/atruskie/bfb7e9ee3df954a29cbc17bdf12405f9). I am opening this PR with their permission.

Notably, this PR only includes a subset of the functionality, specifically the ability to discern type by looking for a key that has the desired type encoded in its value. This PR does not include the functionality to discern type by looking at the presence of unique keys. It does lay the foundation to bring in this additional functionality in the future if desired.

My use case is very similar to the kubernetes use case (where there is a "kind" key), and this PR should satisfy the kubernetes use case.

In addition to less functionality, I have also:
* Removed specific use case types that @atruskie included as examples in the gist.
* Added functionality to specify a max depth/length for buffering values during deserialisation.
* Added a helper method to DeserializerBuilder.

Here is an example of usage:
```csharp
var deserializer = new DeserializerBuilder()
    .WithNamingConvention(CamelCaseNamingConvention.Instance)
    .WithBufferedNodeDeserializer(options => {
        options.AddKeyValueDiscriminator<UserSpecifiedBaseClass>(
            "kind",
            ("deployment", typeof(UserSpecifiedSubclassA)),
            ("service", typeof(UserSpecifiedSubclassB)));
    })
    .Build();
```

I wanted to put this up and gauge interest before working on this more, but the remaining things I know I need to do before merging:
* Comments/code style changes - I have yet to read up on coding conventions but will do
* Tests - I have only tested on net7 currently
* Anything else?

Some other things that could be done in this or future PRs:
* Add back in the other use of buffered deserialisation from @atruskie's gist?
* Handle various case-translations for keys/values?

It is also worth mentioning @EdwardCooke that this also works with the AOT branch we've been testing, it just needs a similar helper method on the `StaticDeserializerBuilder` as that file is not in the codebase yet.

Let me know if this is desired and I'll keep working on it, and thanks for maintaining this library (:

atruskie commented:
Love this; let me know if you want some help to finish it off.

I'd say that the resolver you didn't implement is equally useful and is probably worth adding. I found it more common that yaml files did not include a discriminator key. 

Or if you don't agree, I think it would be better to provide a configuration option method to add a custom ITypeDiscriminator so others can extend as they like.

EdwardCooke commented:
This is great! I've seen a few questions/issues related to this. It'll be nice to have it built in.

A few things I see right off the top that would be great to have addressed
* Comments on public classes and methods
* Adding a way to add a custom `IValueTypeDescriminator` would be a good thing and allow for additional flexibility
* It may be beneficial to mark your methods on your `KeyValueTypeDiscriminator` as virtual so people can use that class and extend it easily.
* It looks like the determined class has to be implementing a base class, and only the inherited class is checked. That may not always be the case. Example use cases where that would not be sufficient.
    * The property being assigned may be an `object` type, not requiring it to be abstract or doing any kind of inheritance.
    * It may also be the class itself, not abstract or inherited.
    * Deep inheritance (class c inherits b, which inherits a and the property type is a).
    * Implementing an interface and the property type is that interface
    * One possible solution would be to have the discriminator do the check to see if it can assign itself to the type, adds more flexibility. And in the default implementation it would call the isassignablefrom method on the type. Pretty sure that's what it is.
 
For the tests, add them to the tests project. It runs them against net47, net 3.1 (for netstandard2.0), net6.0 and net7.0.

JJ11teen commented:
Thanks both, good to know there is appetite for this! 😊 I'll work on the things mentioned and let you know when its ready for a closer look (:

JJ11teen commented:
Okay, I think I have addressed what I consider "design" changes at this point. Still have comments & tests to do though (:
Changes:
* Renamed "IValueTypeDiscriminator" -> "ITypeDiscriminator" and moved into a sub namespace/directory
* Added in @atruskie's other "UniqueKeyTypeDiscriminator" functionality
* Pulled out "TryFindMappingEntry" into existing "ParserExtensions" file
* Added "AddTypeDiscriminator" to Options class to enable users to add their own custom ITypeDiscriminator implementation more easily
* Updated "BufferedNodeDeserializer" to use "IsAssignableFrom", and have removed null & identity type checks - but have kept the concept of a "BaseType" on all discriminators. I believe this supports the scenarios you described @EdwardCooke:
  * The property being assigned may be an object type, not requiring it to be abstract or doing any kind of inheritance.
  * It may also be the class itself, not abstract or inherited.
  * Deep inheritance (class c inherits b, which inherits a and the property type is a).
  * Implementing an interface and the property type is that interface

* However I stopped short of having the discriminator doing the assignment check, as my understanding is that this would result in everything (ie the whole yaml stream) being buffered. The current implementation only starts buffering when the expected type has an associated discriminator. Users can still explicitly use object for the BaseType as you describe, but it is the users responsibility. We can make the BaseType default to object if you think this is a common enough use case and always buffering by default is worth it?

JJ11teen commented:
Actually the BaseType = object scenario is more complex than I previously thought - won't get to it today but will keep thinking (:

JJ11teen commented:
Okay, I have reworked things such that BaseType = object is now supported (see tests).

In order to accomplish this, I have made the `BufferedNodeDeserializer` now wrap the entire list of NodeDeserializers that the Builder has specified. I found this was needed as otherwise the `DictionaryNodeDeserializer` would take precedence when BaseType=object, and the `BufferedNodeDeserializer` was never reached. I also tried just inserting the previous implementation earlier in the builder's deserializer list, but that failed when the deserialized type had a Dictionary as a sub type (as in the case for kubernetes labels/annotations). I'm wondering if you're aware of a better way to handle this NodeDeserializer dependency than what I've got now @EdwardCooke ?

JJ11teen commented:
I can't think of a better solution without changing the signature for `INodeDeserializer`.
```csharp
public interface INodeDeserializer
{
    bool Deserialize(IParser reader, Type expectedType, Func<IParser, Type, object?> nestedObjectDeserializer, out object? value);
}
```

The only output of each `INodeDeserializer` is the deserialized value, (and whether or not the deserializer was able to deserialize any value). This is at odds with the desire for the BufferedNodeDeserializer, which is to "Buffer and deserialize to determine a target type, then reset the buffer and redirect to an appropriate existing INodeDeserializer that matches the newly determined target type".

This is why the BufferedNodeDeserializer needs to wrap all other INodeDeserializers. However all the others also need to exist outside of the BufferedNodeDeserializer for cases when buffering isn't needed/desired.

I don't like the duplication of all of the INodeDeserializers but I think it is a trade off between that and more significantly impacting the existing codebase/functionality/user expectations. Does anyone have any ideas for a more elegant solution? 😂

JJ11teen commented:
Hey @EdwardCooke, just a ping that this is ready for review (:
I rebased on master and added the helper to the static builder yesterday too.

EdwardCooke commented on YamlDotNet/Core/ParserExtensions.cs:
```diff
@@ -147,5 +147,50 @@ public static void SkipThisAndNestedEvents(this IParser parser)
             return Accept<T>(parser, out var _);
         }
+
+        public static bool TryFindMappingEntry(this IParser parser, Func<Scalar, bool> selector, out Scalar? key, out ParsingEvent? value)
```
Can you add some xml comments on here to describe what this method is/does?

EdwardCooke commented on YamlDotNet/Core/ParserExtensions.cs:
```diff
@@ -147,5 +147,50 @@ public static void SkipThisAndNestedEvents(this IParser parser)
+        public static bool TryFindMappingEntry(this IParser parser, Func<Scalar, bool> selector, out Scalar? key, out ParsingEvent? value)
+        {
+            parser.Consume<MappingStart>();
+            do
```
Should this be a `while (parser.Current != null)` to avoid a null reference on your `switch` statement? Not sure if it could be null or if it would break, but it might.

JJ11teen replied:
Makes sense to me (and I've made the change and tests pass), but I'll admit this function is not really one I fully understand. @atruskie if you have time could you double check this logic change and the xml comment I've added?

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/ParserBuffer.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace YamlDotNet.Serialization.BufferedDeserialization
+{
+        public class ParserBuffer : IParser
```
Can you fix the endentation and add some xml comments

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/ParserBuffer.cs:
```diff
@@ -0,0 +1,51 @@
+            current = buffer.First;
+        }
+
+        public ParsingEvent? Current => current?.Value;
```
Can you add xml comments on all of the public members?

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminators/IValueTypeDiscriminator.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace YamlDotNet.Serialization.BufferedDeserialization.TypeDiscriminators
+{
+    public interface ITypeDiscriminator
```
Can you put some xml comments on this?

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminators/KeyValueTypeDiscriminator.cs:
```diff
@@ -0,0 +1,67 @@
+
+        public KeyValueTypeDiscriminator(Type baseType, string targetKey, IDictionary<string, Type> typeMapping)
+        {
+            
```
It's nit picky, but could get rid of the extra whitespace?

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminators/KeyValueTypeDiscriminator.cs:
```diff
@@ -0,0 +1,67 @@
+
+namespace YamlDotNet.Serialization.BufferedDeserialization.TypeDiscriminators
+{
+    public class KeyValueTypeDiscriminator : ITypeDiscriminator
```
Can you add some xml comments on this class and its public members?

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminators/KeyValueTypeDiscriminator.cs:
```diff
@@ -0,0 +1,67 @@
+                }
+                else
+                {
+                    throw new Exception($"Could not determine {BaseType} to deserialize to, {targetKey} has an empty value");
```
Should this case just be ignored? When would it fall in this else statement get used? Throwing an exception here seems like it could be dangerous.

JJ11teen replied:
Yep, makes sense to me to make the discriminator more passive and just ignore things it doesn't understand (:

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminators/KeyValueTypeDiscriminator.cs:
```diff
@@ -0,0 +1,67 @@
+
+            var known = string.Join(",", typeMapping.Keys.ToArray());
+            
+            throw new Exception($"Could not determine {BaseType} to deserialize to, expecting '{targetKey}' to be one of: {known}, but got '{value}'");
```
If instead of throwing an exception, you returned null, here and false above, would you be able to have multiple discriminators triggering on different values of the key?

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminators/UniqueKeyTypeDiscriminator.cs:
```diff
@@ -0,0 +1,44 @@
+            {
+                if (!baseType.IsAssignableFrom(keyValuePair.Value))
+                {
+                    throw new ArgumentOutOfRangeException($"{nameof(typeMapping)} dictionary contains type {keyValuePair.Value} which is not a assignable to {baseType}");
```
Can you add `nameof(typeMapping)` to the arguments on this exception? That way it's automatically included in this exception.

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminators/KeyValueTypeDiscriminator.cs:
```diff
@@ -0,0 +1,67 @@
+            {
+                if (!baseType.IsAssignableFrom(keyValuePair.Value))
+                {
+                    throw new ArgumentOutOfRangeException($"{nameof(typeMapping)} dictionary contains type {keyValuePair.Value} which is not a assignable to {baseType}");
```
Can you add nameof(typeMapping) to the arguments on this exception? That way it's automatically included in this exception.

JJ11teen commented:
Thanks for the review @EdwardCooke! Will aim to get back to this over the coming week 😊

JJ11teen commented:
I think I've addressed everything @EdwardCooke! 😊
@atruskie if you have time it would be great to get your eyes over the change to `TryFindMappingEntry` and how I have described it's functionality in the xml doc. I've given it a go with my understanding but I may have missed a quirk of it's original design/intention.

atruskie commented on YamlDotNet/Core/ParserExtensions.cs:
```diff
@@ -147,5 +147,63 @@ public static void SkipThisAndNestedEvents(this IParser parser)
+                    switch (parser.Current)
+                    {
+                        case Scalar scalar:
+                            // we've found a scalar, check if it's value matches one
+                            // of our predicate
```
```suggestion
                            // we've found a scalar, check if its value matches our predicate
```

atruskie commented on YamlDotNet/Core/ParserExtensions.cs:
```diff
@@ -147,5 +147,63 @@ public static void SkipThisAndNestedEvents(this IParser parser)
         {
             return Accept<T>(parser, out var _);
         }
+
+        /// <summary>
+        /// Checks whether the current event contains a mapping entry that returns true for the specified selector.
+        /// If one is found, it's key is returned as a Scalar and it's value as a ParsingEvent, the mapping entry
+        /// is consumed and the parser is moved to the next event, and the whole function returns true.
+        /// If the event is not a mapping event or a matching key is not found, returns false.
+        /// </summary>
```
```suggestion
         /// Attempts to find a key on a YAML mapping that matches our predicate.
         /// This is useful for scanning a mapping for type discriminator information.
         /// For example: looking for a `kind` key on an object.
         ///
         /// This function only checks mappings, and only looks at the current depth.
         ///
         /// If the event is a mapping and has a key that satisfies the predicate the scan 
         /// will stop, return true, and set
         /// <paramref>key</paramref> and <paramref>value</paramref>.
         ///
        /// If the event is not a mapping event or a matching key is not found, returns false.
```

Your description is correct but perhaps a little jargony. I've added my suggestion, feel free to adapt how you like.

atruskie commented on YamlDotNet/Serialization/BufferedDeserialization/BufferedNodeDeserializer.cs:
```diff
@@ -0,0 +1,99 @@
+    /// Usually you will want all default <see cref="INodeDeserializer" />s that exist in the outer
+    /// <see cref="Deserializer" /> to also be used as inner deserializers.
+    /// </summary>
+    public class BufferedNodeDeserializer : INodeDeserializer
```
Perhaps this type name should be changed to `TypeDiscriminatingNodeDeserializer`. 

It reflects its actual purpose (discriminating types) rather than describing its method of action (buffering).

atruskie reviewed (commented):
This is fantastic!

A few notes:

- Your tests don't cover a fall through case (i.e. two registrations of `AddKeyValueTypeDiscriminator` or `AddUniqueKeyTypeDiscriminator`) but the code seems to support it. I.e. would this work:

```
            var bufferedDeserializer = new DeserializerBuilder()
                .WithNamingConvention(CamelCaseNamingConvention.Instance)
                .WithBufferedNodeDeserializer(options => {
                    options.AddKeyValueTypeDiscriminator<KubernetesResource>(
                        "kind",
                        new Dictionary<string, Type>()
                        {
                            { "Namespace", typeof(KubernetesNamespace) },
                        })
                       .AddKeyValueTypeDiscriminator<KubernetesResource>(
                        "kind",
                        new Dictionary<string, Type>()
                        {
                            { "Service", typeof(KubernetesService) }
                        });
                    },
                    maxDepth: 3,
                    maxLength: 40)
                .Build();
```

- The name `BufferedNodeDeserializer` is probably one I would change (I know I chose the name originally, my bad) since it doesn't describe what it's doing but rather how - and how is unimportant. Suggestion: `TypeDiscriminatingNodeDeserializer`.

I'm so hyped you did the work for this. There's no way I would have had the capacity, so thank you!

JJ11teen commented:
Thanks @atruskie! Have made those changes (:

JJ11teen commented:
Rebased onto master again (:

EdwardCooke commented on YamlDotNet/Serialization/DeserializerBuilder.cs:
```diff
@@ -246,6 +247,28 @@ public DeserializerBuilder WithoutNodeDeserializer(Type nodeDeserializerType)
+            configureTypeDiscriminatingNodeDeserializerOptions(options);
+            // We use all current NodeDeserializers as the inner deserializers for the TypeDiscriminatingNodeDeserializer,
+            // so that it can successfully deserialize anything our root deserializer can.
+            var TypeDiscriminatingNodeDeserializer = new TypeDiscriminatingNodeDeserializer(nodeDeserializerFactories.BuildComponentList(), options.discriminators, maxDepth, maxLength);
```
This should be a lower case `T` on the variable name.

EdwardCooke commented on YamlDotNet/Serialization/StaticDeserializerBuilder.cs:
```diff
@@ -205,6 +206,28 @@ public StaticDeserializerBuilder WithoutNodeDeserializer(Type nodeDeserializerTy
+            configureTypeDiscriminatingNodeDeserializerOptions(options);
+            // We use all current NodeDeserializers as the inner deserializers for the TypeDiscriminatingNodeDeserializer,
+            // so that it can successfully deserialize anything our root deserializer can.
+            var TypeDiscriminatingNodeDeserializer = new TypeDiscriminatingNodeDeserializer(nodeDeserializerFactories.BuildComponentList(), options.discriminators, maxDepth, maxLength);
```
Lowercase `T` here as well.

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminatingNodeDeserializer.cs:
```diff
@@ -0,0 +1,99 @@
+using System;
```
Missing required file header

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminators/UniqueKeyTypeDiscriminator.cs:
```diff
@@ -0,0 +1,67 @@
+using System;
```
Missing required file header

EdwardCooke commented on YamlDotNet.Test/Serialization/BufferedDeserialization/KeyValueTypeDiscriminatorTests.cs:
```diff
@@ -0,0 +1,244 @@
+using System;
```
Missing required file header

EdwardCooke commented on YamlDotNet.Test/Serialization/BufferedDeserialization/TypeDiscriminatingNodeDeserializerTests.cs:
```diff
@@ -0,0 +1,68 @@
+using System;
```
Missing required file header

EdwardCooke commented on YamlDotNet.Test/Serialization/BufferedDeserialization/UniqueKeyTypeDiscriminatorTests.cs:
```diff
@@ -0,0 +1,82 @@
+using System;
```
Missing required file header

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/ITypeDiscriminatingNodeDeserializerOptions.cs:
```diff
@@ -0,0 +1,26 @@
+using System;
```
Missing required file header

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/ParserBuffer.cs:
```diff
@@ -0,0 +1,71 @@
+using System;
```
Missing required file header

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminatingNodeDeserializerOptions.cs:
```diff
@@ -0,0 +1,46 @@
+using System;
```
Missing required file header

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminators/IValueTypeDiscriminator.cs:
```diff
@@ -0,0 +1,30 @@
+using System;
```
Missing required file header

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminators/KeyValueTypeDiscriminator.cs:
```diff
@@ -0,0 +1,75 @@
+using System;
```
Missing required file header

EdwardCooke commented on YamlDotNet/Serialization/DeserializerBuilder.cs:
```diff
@@ -25,6 +25,7 @@
 using System.Diagnostics.CodeAnalysis;
```
Missing required file header

JJ11teen replied:
This one already exists 😅 Thanks for going through and being explicit with what you want though, I appreciate it 😊

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminatingNodeDeserializer.cs:
```diff
@@ -0,0 +1,99 @@
+    /// a dotnet output type for the yaml node. As the node is buffered, the <see cref="ITypeDiscriminator" />s are
+    /// able to examine the actual values within, and use these when discriminating a type.
+    /// Once a matching type is found, the TypeDiscriminatingNodeDeserializer uses it's inner deserializers to perform
+    /// the final deserialization for that type & object.
```
This should be `&amp;` and not just an `&` sign

EdwardCooke commented on YamlDotNet/Serialization/BufferedDeserialization/TypeDiscriminatingNodeDeserializer.cs:
```diff
@@ -0,0 +1,99 @@
+            value = null;
+            return false;
+        }
+    }   
```
I know its a little nit picky, but it throws a warning on build because of the extra whitespace after the curly brace here.

EdwardCooke commented on YamlDotNet/Core/ParserExtensions.cs:
```diff
@@ -147,5 +147,70 @@ public static void SkipThisAndNestedEvents(this IParser parser)
+        /// <param name="value">The matching value of the mapping as a ParsingEvent, or null if no matching key found</param>
+        /// <returns>Returns true if the current event is a mapping entry with a key that matches the selector;
+        /// otherwise returns false.</returns>
+        public static bool TryFindMappingEntry(this IParser parser, Func<Scalar, bool> selector, [MaybeNullWhen(false)] out Scalar? key, [MaybeNullWhen(false)] out ParsingEvent? value)
```
The `parser` parameter here needs an xml comment above.

EdwardCooke commented:
Once the latest comments are fixed we should be good to merge this in.

JJ11teen commented:
All done @EdwardCooke! (:

EdwardCooke commented:
I’ll merge this in when I get time, should be this week.

pmikstacki commented:
Hi! When it will be available as nuget package?

aaubry commented:
This feature has been released in [version 13.1.0](https://github.com/aaubry/YamlDotNet/releases/tag/v13.1.0).

rcdailey commented:
I have just pulled down the 13.1.0 release and tried using this, but the lack of documentation is killing me.

- This PR didn't serve as a helpful form of documentation due to the change in class names over time
- There's no documentation markdown anywhere that has tutorials or the like to help people get started with this.
- Some public interfaces, like `ITypeDiscriminatingNodeDeserializerOptions`, nave no C# documentation so it's unclear, for example, what the difference is between `AddUniqueKeyTypeDiscriminator` and `AddKeyValueTypeDiscriminator`.

Don't get me wrong, I'm not here just to complain. Just pointing out something that I think will make it difficult for users to adopt this new feature. I'm sure if I swim into the unit tests I'll eventually figure it out, but hours are pretty valuable these days...

In any case, I appreciate the months you folks spent on getting this ready.

EdwardCooke commented:
Thanks for the feedback, I'll work on building out some documentation in the wiki, I know this particular feature was a bit complex in nature. I'll also work on getting those missing XML comments on the interface.

EdwardCooke commented:
Here's a little wiki page documenting how to use this new feature

https://github.com/aaubry/YamlDotNet/wiki/Deserialization---Type-Discrimators

R3ven6eR reviewed (commented):
- [ ]

rcdailey commented:
> Here's a little wiki page documenting how to use this new feature
> 
> [aaubry/YamlDotNet/wiki/Deserialization---Type-Discrimators](https://github.com/aaubry/YamlDotNet/wiki/Deserialization---Type-Discrimators)

For future visitors, this is the correct link: https://github.com/aaubry/YamlDotNet/wiki/Deserialization---Type-Discriminators

Thanks again for the examples!
